File: backend/charmap.py
# Vestaboard-compatible character set
# Index 0 = blank, 1-26 = A-Z, 27-36 = 1-0, 37+ = symbols, 71-77 = colors

CHARS = [
    ' ',  # 0  blank
    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J',  # 1-10
    'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',  # 11-20
    'U', 'V', 'W', 'X', 'Y', 'Z',                        # 21-26
    '1', '2', '3', '4', '5', '6', '7', '8', '9', '0',   # 27-36
    '!', '@', '#', '$', '(', ')', '-', '+', '&', '=',    # 37-46
    ';', ':', "'", '"', '%', ',', '.', '/', '?', '°',    # 47-56
    '♥', '♦', '♣', '♠',               # 57-60 hearts/diamonds/clubs/spades
    '★', '☆',                                    # 61-62 star/outline star
    '←', '↑', '→', '↓',               # 63-66 arrows
    '·',                                              # 67 middle dot
    '■',                                              # 68 filled square
    '○',                                              # 69 circle
    ' ',                                                   # 70 reserved
    'RED', 'ORANGE', 'YELLOW', 'GREEN', 'BLUE', 'VIOLET', 'WHITE',  # 71-77 color tiles
]

CHAR_TO_CODE = {}
for i, ch in enumerate(CHARS):
    if ch not in CHAR_TO_CODE and ch != ' ' or i == 0:
        CHAR_TO_CODE[ch] = i

# Additional mappings for common chars
CHAR_TO_CODE.update({
    ' ': 0,
    '\n': 0,
    '\t': 0,
})

COLOR_CODES = {
    'RED': 71, 'ORANGE': 72, 'YELLOW': 73,
    'GREEN': 74, 'BLUE': 75, 'VIOLET': 76, 'WHITE': 77,
}

COLOR_HEX = {
    71: '#e63946',  # red
    72: '#f4a261',  # orange
    73: '#e9c46a',  # yellow
    74: '#2a9d8f',  # green
    75: '#457b9d',  # blue
    76: '#7b2d8b',  # violet
    77: '#f1faee',  # white
}


def char_to_code(ch: str) -> int:
    ch_upper = ch.upper()
    if ch_upper in CHAR_TO_CODE:
        return CHAR_TO_CODE[ch_upper]
    return 0  # unknown -> blank


def text_to_row(text: str, cols: int) -> list[int]:
    row = []
    for ch in text.upper():
        row.append(char_to_code(ch))
        if len(row) >= cols:
            break
    while len(row) < cols:
        row.append(0)
    return row


def text_to_matrix(text: str, rows: int, cols: int) -> list[list[int]]:
    """Word-wrap text into a rows×cols matrix."""
    words = text.upper().split()
    lines = []
    current = ""
    for word in words:
        if len(word) > cols:
            # Split long word
            if current:
                lines.append(current)
                current = ""
            while word:
                lines.append(word[:cols])
                word = word[cols:]
        elif len(current) + len(word) + (1 if current else 0) <= cols:
            current = (current + " " + word).strip() if current else word
        else:
            lines.append(current)
            current = word
    if current:
        lines.append(current)

    # Truncate to available rows
    lines = lines[:rows]

    # Center each line horizontally and convert to codes
    text_rows = []
    for line in lines:
        padding = (cols - len(line)) // 2
        centered = " " * padding + line
        text_rows.append(text_to_row(centered, cols))

    # Vertical centering: blank rows above so text block is centered
    top_pad = (rows - len(text_rows)) // 2
    matrix = [[0] * cols for _ in range(top_pad)]
    matrix.extend(text_rows)
    while len(matrix) < rows:
        matrix.append([0] * cols)

    return matrix


def blank_matrix(rows: int, cols: int) -> list[list[int]]:
    return [[0] * cols for _ in range(rows)]


# ── Colored text markup ───────────────────────────────────────────────────────
# {red}HAPPY HOUR{/} — colors the enclosed characters. Something a physical
# split-flap board can't do.

import re as _re

MARKUP_COLORS = {
    "red": "#e63946", "orange": "#f4a261", "yellow": "#e9c46a",
    "green": "#2a9d8f", "blue": "#457b9d", "violet": "#7b2d8b",
    "white": "#f1faee",
}

_MARKUP_RE = _re.compile(r"\{(red|orange|yellow|green|blue|violet|white|/)\}", _re.IGNORECASE)


def parse_colored_text(text: str) -> tuple[str, list]:
    """Strip {color}…{/} markup. Returns (clean_text, per-char hex-or-None)."""
    clean_parts: list[str] = []
    colors: list = []
    current = None
    pos = 0
    for m in _MARKUP_RE.finditer(text):
        seg = text[pos:m.start()]
        clean_parts.append(seg)
        colors.extend([current] * len(seg))
        tag = m.group(1).lower()
        current = None if tag == "/" else MARKUP_COLORS[tag]
        pos = m.end()
    seg = text[pos:]
    clean_parts.append(seg)
    colors.extend([current] * len(seg))
    return "".join(clean_parts), colors


def text_to_matrix_colored(text: str, rows: int, cols: int):
    """Like text_to_matrix but honoring color markup. Returns
    (matrix, color_map) — color_map is rows×cols of hex-or-None, or None
    when the text carries no markup."""
    clean, char_colors = parse_colored_text(text)

    # Words as (char, color) pair lists, mirroring text_to_matrix's wrapping
    words: list[list] = []
    current_word: list = []
    for ch, color in zip(clean, char_colors, strict=True):
        if ch.isspace():
            if current_word:
                words.append(current_word)
                current_word = []
        else:
            current_word.append((ch.upper(), color))
    if current_word:
        words.append(current_word)

    lines: list[list] = []
    line: list = []
    for word in words:
        if len(word) > cols:
            if line:
                lines.append(line)
                line = []
            while word:
                lines.append(word[:cols])
                word = word[cols:]
        elif len(line) + len(word) + (1 if line else 0) <= cols:
            if line:
                line.append((" ", None))
            line.extend(word)
        else:
            lines.append(line)
            line = word
    if line:
        lines.append(line)
    lines = lines[:rows]

    matrix_rows, color_rows = [], []
    for ln in lines:
        pad = max(0, (cols - len(ln)) // 2)
        row = [0] * cols
        crow = [None] * cols
        for i, (ch, color) in enumerate(ln):
            if pad + i < cols:
                row[pad + i] = char_to_code(ch)
                crow[pad + i] = color
        matrix_rows.append(row)
        color_rows.append(crow)

    top_pad = max(0, (rows - len(matrix_rows)) // 2)
    matrix = [[0] * cols for _ in range(top_pad)]
    cmap = [[None] * cols for _ in range(top_pad)]
    matrix.extend(matrix_rows)
    cmap.extend(color_rows)
    while len(matrix) < rows:
        matrix.append([0] * cols)
        cmap.append([None] * cols)

    if all(c is None for r in cmap for c in r):
        cmap = None
    return matrix, cmap

File: frontend/src/utils/charmap.js
// Mirrors backend charmap.py — Vestaboard-compatible character set
export const CHARS = [
  ' ',                                                          // 0
  'A','B','C','D','E','F','G','H','I','J',                     // 1-10
  'K','L','M','N','O','P','Q','R','S','T',                     // 11-20
  'U','V','W','X','Y','Z',                                     // 21-26
  '1','2','3','4','5','6','7','8','9','0',                     // 27-36
  '!','@','#','$','(',')','-','+','&','=',                     // 37-46
  ';',':',`'`,'"','%',',','.','/',`?`,'°',                    // 47-56
  '♥','♦','♣','♠',                                            // 57-60
  '★','☆',                                                    // 61-62
  '←','↑','→','↓',                                           // 63-66
  '·','■','○',                                                // 67-69
  ' ',                                                         // 70 reserved
  'RED','ORANGE','YELLOW','GREEN','BLUE','VIOLET','WHITE',     // 71-77 color tiles
]

export const COLOR_HEX = {
  71: '#e63946',
  72: '#f4a261',
  73: '#e9c46a',
  74: '#2a9d8f',
  75: '#457b9d',
  76: '#7b2d8b',
  77: '#f1faee',
}

export const COLOR_NAMES = {
  71: 'RED', 72: 'ORANGE', 73: 'YELLOW',
  74: 'GREEN', 75: 'BLUE', 76: 'VIOLET', 77: 'WHITE',
}

export const isColorCode = (code) => code >= 71 && code <= 77

export const codeToChar = (code) => {
  if (code < 0 || code >= CHARS.length) return ' '
  return CHARS[code]
}

const CHAR_TO_CODE = {}
CHARS.forEach((ch, i) => {
  if (!(ch in CHAR_TO_CODE) || i === 0) CHAR_TO_CODE[ch] = i
})
CHAR_TO_CODE[' '] = 0

export const charToCode = (ch) => CHAR_TO_CODE[ch.toUpperCase()] ?? 0

// ── Text sanitizing — mirrors backend charmap.sanitize_text ──────────────────
// Fold "smart" punctuation and accented letters to ASCII look-alikes and drop
// anything still unrenderable, so text collapses (JOSÉ'S → JOSE'S) instead of
// leaving a blank tile mid-word.
const PUNCT_FOLD = {
  '‘': "'", '’': "'", '‚': "'", '′': "'", '`': "'",
  '“': '"', '”': '"', '„': '"', '″': '"',
  '–': '-', '—': '-', '―': '-', '−': '-',
  '…': '...',
  ' ': ' ', ' ': ' ', ' ': ' ', '​': '',
  '•': '·', '·': '·',
}

export function sanitizeText(text) {
  if (!text) return text ?? ''
  let out = ''
  for (const ch of text) {
    if (ch in PUNCT_FOLD) { out += PUNCT_FOLD[ch]; continue }
    if (ch === '\n' || ch === '\t' || ch.toUpperCase() in CHAR_TO_CODE) { out += ch; continue }
    // Fold accents to base letters, then keep only renderable pieces
    const base = ch.normalize('NFKD').replace(/\p{M}/gu, '')
    for (const b of base) if (b.toUpperCase() in CHAR_TO_CODE) out += b
  }
  return out
}

/** Mirrors backend charmap.text_to_matrix: word-wrap, center each line
 * horizontally, center the block vertically. Used for local previews. */
export function textToMatrix(text, rows, cols) {
  const words = sanitizeText(text).toUpperCase().split(/\s+/).filter(Boolean)
  const lines = []
  let current = ''
  for (let word of words) {
    if (word.length > cols) {
      if (current) { lines.push(current); current = '' }
      while (word) { lines.push(word.slice(0, cols)); word = word.slice(cols) }
    } else if (current.length + word.length + (current ? 1 : 0) <= cols) {
      current = current ? `${current} ${word}` : word
    } else {
      lines.push(current)
      current = word
    }
  }
  if (current) lines.push(current)

  const textRows = lines.slice(0, rows).map(line => {
    const pad = Math.max(0, Math.floor((cols - line.length) / 2))
    const centered = ' '.repeat(pad) + line
    const row = []
    for (let c = 0; c < cols; c++) row.push(charToCode(centered[c] ?? ' '))
    return row
  })

  const topPad = Math.max(0, Math.floor((rows - textRows.length) / 2))
  const matrix = Array.from({ length: topPad }, () => Array(cols).fill(0))
  matrix.push(...textRows)
  while (matrix.length < rows) matrix.push(Array(cols).fill(0))
  return matrix
}

// ── Colored text markup — mirrors backend charmap.parse_colored_text ─────────
// {red}HAPPY HOUR{/} colors the enclosed characters.

export const MARKUP_COLORS = {
  red: '#e63946', orange: '#f4a261', yellow: '#e9c46a',
  green: '#2a9d8f', blue: '#457b9d', violet: '#7b2d8b',
  white: '#f1faee',
}

const MARKUP_RE = /\{(red|orange|yellow|green|blue|violet|white|\/)\}/gi

/** Strip {color}…{/} markup. Returns [cleanText, perCharHexOrNull[]]. */
export function parseColoredText(text) {
  const cleanParts = []
  const colors = []
  let current = null
  let pos = 0
  MARKUP_RE.lastIndex = 0
  let m
  while ((m = MARKUP_RE.exec(text)) !== null) {
    const seg = text.slice(pos, m.index)
    cleanParts.push(seg)
    for (let i = 0; i < seg.length; i++) colors.push(current)
    const tag = m[1].toLowerCase()
    current = tag === '/' ? null : MARKUP_COLORS[tag]
    pos = m.index + m[0].length
  }
  const seg = text.slice(pos)
  cleanParts.push(seg)
  for (let i = 0; i < seg.length; i++) colors.push(current)
  return [cleanParts.join(''), colors]
}

/** Like textToMatrix but honoring color markup. Returns [matrix, colorMap]
 * — colorMap is rows×cols of hex-or-null, or null when no markup. */
export function textToMatrixColored(text, rows, cols) {
  const [clean, charColors] = parseColoredText(text)

  // Fold/strip unrenderable characters, keeping each survivor's color, then
  // split into words — mirrors textToMatrix's wrapping. A char may vanish
  // (emoji) or expand (… → ...); colors follow one-to-one.
  const words = []
  let currentWord = []
  for (let i = 0; i < clean.length; i++) {
    for (const c of sanitizeText(clean[i])) {
      if (/\s/.test(c)) {
        if (currentWord.length) { words.push(currentWord); currentWord = [] }
      } else {
        currentWord.push([c.toUpperCase(), charColors[i]])
      }
    }
  }
  if (currentWord.length) words.push(currentWord)

  const lines = []
  let line = []
  for (let word of words) {
    if (word.length > cols) {
      if (line.length) { lines.push(line); line = [] }
      while (word.length) { lines.push(word.slice(0, cols)); word = word.slice(cols) }
    } else if (line.length + word.length + (line.length ? 1 : 0) <= cols) {
      if (line.length) line.push([' ', null])
      line.push(...word)
    } else {
      lines.push(line)
      line = word
    }
  }
  if (line.length) lines.push(line)

  const matrixRows = []
  const colorRows = []
  for (const ln of lines.slice(0, rows)) {
    const pad = Math.max(0, Math.floor((cols - ln.length) / 2))
    const row = Array(cols).fill(0)
    const crow = Array(cols).fill(null)
    ln.forEach(([ch, color], i) => {
      if (pad + i < cols) {
        row[pad + i] = charToCode(ch)
        crow[pad + i] = color
      }
    })
    matrixRows.push(row)
    colorRows.push(crow)
  }

  const topPad = Math.max(0, Math.floor((rows - matrixRows.length) / 2))
  const matrix = Array.from({ length: topPad }, () => Array(cols).fill(0))
  const cmap = Array.from({ length: topPad }, () => Array(cols).fill(null))
  matrix.push(...matrixRows)
  cmap.push(...colorRows)
  while (matrix.length < rows) {
    matrix.push(Array(cols).fill(0))
    cmap.push(Array(cols).fill(null))
  }

  const hasColor = cmap.some(r => r.some(c => c !== null))
  return [matrix, hasColor ? cmap : null]
}
